web/pandas/pdeps/0009-io-extensions.md

```diff
@@ -0,0 +1,395 @@
+pandas dataframes as a possible interface for connectors.
+
+In case a `read` method returned something different than a PyArrow table,
+pandas would raise an exception. pandas would expect all `to_` methods to have
+`table: pyarrow.Table | pandas.DataFrame` as the first parameter, and it would
+raise an exception otherwise. The `table` parameter would be exposed as the
+`self` parameter of the `to_*` method in pandas.
+
+In case the Apache Arrow interface is used, metadata not supported by Apache
+Arrow may be provided by users. For example the column to use for row indices
+or the data type backend to use in the object being created. This would be
+managed independently from the connectors. Given the previous example, a new
+argument `index_col` could be added directly into pandas to the function or
+method automatically generated from the entrypoint. Since this would apply to
+all functions and methods automatically generated, it would also improve the
+consistency of pandas connectors. For example:
+
+```python
+def read_myformat(index_col=None, *args, **kwargs):
+    # The third-party connector doesn't need to know about functionality
+    # specific to pandas like the row index
+    arrow_table = third_party_connector.from_myformat(*args, **kwargs)
+
+    df = convert_arrow_table_to_pandas_dataframe(arrow_table)
+
+    # Transformations to the dataframe with custom parameters is possible
+    if index_col is not None:
+        df = df.set_index(index_col)
+
+    return df
+```
+
+#### Connector guidelines
+
+In order to provide a better and more consistent experience to users, guidelines
+will be created to unify terminology and behavior. Some of the topics to unify are
+defined next.
+
+**Existence and naming of columns**, since many connectors are likely to provide
+similar features, like loading only a subset of columns in the data, or dealing
+with file names. Examples of recommendations to connector developers:
+
+- `columns`: Use this argument to let the user load a subset of columns. Allow a
+  list or tuple.
+- `path`: Use this argument if the dataset is a file in the file disk. Allow a string,
+  a `pathlib.Path` object, or a file descriptor. For a string object, allow URLs that
+  will be automatically download, compressed files that will be automatically
+  uncompressed, etc. A library can be provided to deal with those in an easier and
+  more consistent way.
+- `schema`: For datasets that don't have a schema (e.g. `csv`), allow providing an
+  Apache Arrow schema instance, and automatically infer types if not provided.
+
+Note that the above are only examples of guidelines for illustration, and not
+a proposal of the guidelines.
+
+**Guidelines to avoid name conflicts**. Since it is expected that more than one
+implementation exists for certain formats, as it already happens, guidelines on
+how to name connectors would be created. The easiest approach is probably to use
```

Comment:
Is it possible for pandas to raise if it detects that 2 entrypoints are trying to register the same named `read_` and `to_` function? If so, I am not sure if it's necessary to enforce naming conventions necessarily

Reply:
Yes, fully agree, I thought it was in the proposal that pandas would raise if more than one connector with the same name exists, but seems like it's not there, I can add it.

But I do think some naming guidelines could still be helpful. Like recommending to use something like `read_csv_strict` instead of `read_strict_csv`... Or whatever we want to recommend that helps consistency.